File: companies/Murex/90.subsets-ii.java
import java.util.*;

/**
 * Algorithm:
 * Sort so duplicates are adjacent. During backtracking, skip nums[i] when it is
 * equal to nums[i - 1] and i is not the first choice at this recursion depth.
 *
 * Complexity:
 * Time O(U * n), where U is the number of unique subsets. Space O(n) recursion
 * excluding output.
 */
class Solution {
    private int[] nums;
    private List<Integer> path;
    private List<List<Integer>> ans;

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        this.nums = nums;
        path = new ArrayList<>();
        ans = new ArrayList<>();
        dfs(0);
        return ans;
    }

    private void dfs(int start) {
        ans.add(new ArrayList<>(path));
        for (int i = start; i < nums.length; i++) {
            if (i > start && nums[i] == nums[i - 1]) {
                continue;
            }
            path.add(nums[i]);
            dfs(i + 1);
            path.remove(path.size() - 1);
        }
    }
}

File: data structure and algorithms/Bit Manipulation/90.subsets-ii.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        /*
        Approach:
        Sort so duplicates are adjacent. During backtracking, skip a duplicate
        if the previous equal value was not chosen at this same recursion depth;
        otherwise the same subset would be generated twice.

        Complexity: O(U*n) time where U is the number of unique subsets, O(n)
        recursion space excluding output.
        */
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        vector<int> path;
        function<void(int)> dfs = [&](int start) {
            ans.push_back(path);
            for (int i = start; i < (int)nums.size(); ++i) {
                if (i > start && nums[i] == nums[i - 1]) continue;
                path.push_back(nums[i]);
                dfs(i + 1);
                path.pop_back();
            }
        };
        dfs(0);
        return ans;
    }
};
